test/foundry/mocks/libraries/HederaTokenValidation.sol

```diff
@@ -0,0 +1,333 @@
+
+            if (partialNonFungibleTokenInfo.ownerId != expectedOwner) {
+                return (false, HederaResponseCodes.SENDER_DOES_NOT_OWN_NFT_SERIAL_NO);
+            }
+        }
+
+        success = true;
+        responseCode = HederaResponseCodes.SUCCESS;
+    }
+
+    function _validateFungibleBalance(
+        address token,
+        address owner,
+        uint amount,
+        mapping(address => bool) storage _isFungible
+    ) internal view returns (bool success, int64 responseCode) {
+        if (_isFungible[token]) {
+            HederaFungibleToken hederaFungibleToken = HederaFungibleToken(token);
+
+            bool sufficientBalance = hederaFungibleToken.balanceOf(owner) >= uint64(amount);
+
+            if (!sufficientBalance) {
+                return (false, HederaResponseCodes.INSUFFICIENT_TOKEN_BALANCE);
+            }
+        }
+
+        success = true;
+        responseCode = HederaResponseCodes.SUCCESS;
+    }
+
+    function _validateTokenSufficiency(
+        address token,
+        address owner,
+        int64 amount,
+        int64 serialNumber,
+        mapping(address => bool) storage _isFungible,
+        mapping(address => bool) storage _isNonFungible,
+        mapping(address => mapping(int64 => IHtsPrecompileMock.PartialNonFungibleTokenInfo)) storage _partialNonFungibleTokenInfos
+    ) internal view returns (bool success, int64 responseCode) {
+
+        uint256 amountU256 = uint64(amount);
+        uint256 serialNumberU256 = uint64(serialNumber);
+        return _validateTokenSufficiency(token, owner, amountU256, serialNumberU256, _isFungible, _isNonFungible, _partialNonFungibleTokenInfos);
+    }
+
+    function _validateTokenSufficiency(
+        address token,
+        address owner,
+        uint256 amount,
+        uint256 serialNumber,
+        mapping(address => bool) storage _isFungible,
+        mapping(address => bool) storage _isNonFungible,
+        mapping(address => mapping(int64 => IHtsPrecompileMock.PartialNonFungibleTokenInfo)) storage _partialNonFungibleTokenInfos
+    ) internal view returns (bool success, int64 responseCode) {
+
+        if (_isFungible[token]) {
+            return _validateFungibleBalance(token, owner, amount, _isFungible);
+        }
+
+        if (_isNonFungible[token]) {
+            return _validateNftOwnership(token, owner, serialNumber, _isNonFungible, _partialNonFungibleTokenInfos);
+        }
+    }
+
+    function _validateFungibleApproval(
+        address token,
+        address spender,
+        address from,
+        uint256 amount,
+        mapping(address => bool) storage _isFungible
+    ) internal view returns (bool success, int64 responseCode) {
+        if (_isFungible[token]) {
+
+            uint256 allowance = HederaFungibleToken(token).allowance(from, spender);
+
+            // TODO: do validation for other allowance response codes such as SPENDER_DOES_NOT_HAVE_ALLOWANCE and MAX_ALLOWANCES_EXCEEDED
```

Comment:
Just open a ticket to capture this. You can make code comment the ticket

Reply:
I opened a few issues to capture all the TODOs, but I think it would be best to leave the TODO comments for the developer working on the issue to more easily understand where the issue is.

Reply:
That's fine, no biggie